src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp

```diff
@@ -0,0 +1,342 @@
+                     const double b_squared,
+                     const double r_dot_b_squared) noexcept
+      : v_squared_target_(1.0 - 1.0 / square(w_target)),
+        r_squared_(r_squared),
+        b_squared_(b_squared),
+        r_dot_b_squared_(r_dot_b_squared) {}
+
+  double operator()(const double mu) const noexcept {
+    const double x = compute_x(mu, b_squared_);
+    const double r_bar_squared =
+        compute_r_bar_squared(mu, x, r_squared_, r_dot_b_squared_);
+    // v = mu*r_bar, see text after Equation (31)
+    // target is v satisfying W(mu) = D/ rho_{min/max}
+    return square(mu) * r_bar_squared - v_squared_target_;
+  }
+
+ private:
+  const double v_squared_target_;
+  const double r_squared_;
+  const double b_squared_;
+  const double r_dot_b_squared_;
+};
+
+// Function whose root is upper bracket of master function, see Sec. II.F
+class AuxiliaryFunction {
+ public:
+  AuxiliaryFunction(const double h_0, const double r_squared,
+                    const double b_squared,
+                    const double r_dot_b_squared) noexcept
+      : h_0_(h_0),
+        r_squared_(r_squared),
+        b_squared_(b_squared),
+        r_dot_b_squared_(r_dot_b_squared) {}
+
+  double operator()(const double mu) const noexcept {
+    const double x = compute_x(mu, b_squared_);
+    const double r_bar_squared =
+        compute_r_bar_squared(mu, x, r_squared_, r_dot_b_squared_);
+    // Equation (49)
+    return mu * sqrt(square(h_0_) + r_bar_squared) - 1.0;
+  }
+
+ private:
+  const double h_0_;
+  const double r_squared_;
+  const double b_squared_;
+  const double r_dot_b_squared_;
+};
+
+// Master function, see Sec. II.E
+template <size_t ThermodynamicDim>
+class FunctionOfMu {
+ public:
+  FunctionOfMu(const double total_energy_density,
+               const double momentum_density_squared,
+               const double momentum_density_dot_magnetic_field,
+               const double magnetic_field_squared,
+               const double rest_mass_density_times_lorentz_factor,
+               const EquationsOfState::EquationOfState<true, ThermodynamicDim>&
+                   equation_of_state) noexcept
+      : q_(total_energy_density / rest_mass_density_times_lorentz_factor - 1.0),
+        r_squared_(momentum_density_squared /
+                   square(rest_mass_density_times_lorentz_factor)),
+        b_squared_(magnetic_field_squared /
+                   rest_mass_density_times_lorentz_factor),
+        r_dot_b_squared_(square(momentum_density_dot_magnetic_field) /
+                         cube(rest_mass_density_times_lorentz_factor)),
+        rest_mass_density_times_lorentz_factor_(
+            rest_mass_density_times_lorentz_factor),
+        equation_of_state_(equation_of_state),
+        h_0_(equation_of_state_.specific_enthalpy_lower_bound()),
+        v_0_squared_(compute_v_0_squared(r_squared_, h_0_)) {}
+
+  std::pair<double, double> root_bracket(
```

Comment:
can you define these functions of out line? The struct is a bit difficult to read